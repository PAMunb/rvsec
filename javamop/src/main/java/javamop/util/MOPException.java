// Copyright (c) 2002-2014 JavaMOP Team. All Rights Reserved.
package javamop.util;

/**
 * A general exception class signifying something went wrong in the JavaMOP logic.
 * @author fengchen
 */
public class MOPException extends Exception {
    private static final long serialVersionUID = 2145299315023315212L;
    
    /**
     * Wrap another exception as a MOP logic exception.
     * @param e The exception to wrap.
     */
    public MOPException(final Exception e){
        super("MOP Expection:" + e.getMessage(), e);
    }
    
    /**
     * Construct a MOPException with an informative message.
     * @param str An informative message describing the error.
     */
    public MOPException(final String str){
        super(str);
    }
    
    /**
     * Wrap another exception as a MOP logic exception, with an additional message.
     * @param message A descriptive message for the error.
     * @param e The exception to be wrapped.
     */
    public MOPException(final String message, final Exception e) {
        super("MOP Exception: " + message + ": " + e.getMessage(), e);
    }
}
