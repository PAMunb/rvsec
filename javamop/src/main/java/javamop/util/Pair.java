// Copyright (c) 2002-2014 JavaMOP Team. All Rights Reserved.
package javamop.util;

/**
 * A generic immutable Pair class for holding two values of any type.
 */
public class Pair<A, B>{
    private final A left;
    private final B right;
    
    /**
     * Construct the pair.
     * @param a The left side of the pair.
     * @param b The right side of the pair.
     */
    public Pair(final A a, final B b){
        this.left = a;
        this.right = b;
    }
    
    /**
     * Retrieve the left side of the pair.
     * @return The left side.
     */
    public A getLeft() {
        return left;
    }
    
    /**
     * Retrieve the right side of the pair.
     * @return The right side
     */
    public B getRight() {
        return right;
    }
}
