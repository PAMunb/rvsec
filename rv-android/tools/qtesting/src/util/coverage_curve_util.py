import os, shutil, xml, xml.dom.minidom, re, collections, configparser, sys
APP_SOURCE_FILE = ''
APK_NAME = ''
Benchmark = ''
MANIFEST_FILE = ''
TARGET_BASED_COVERAGE_FILE = ''
ORIGINAL_COVERAGE_FILE = ''
NEW_COVERAGE_FILE = ''
TIME_INTERVAL = 30
MAX_ID = 0

def copy_restart_coverage_file(target_based_coverage_file, original_coverage_file):
    file_list = os.listdir(target_based_coverage_file)
    if not file_list:
        return
    for f in file_list:
        if '.ec' in f:
            shutil.copy2(os.path.join(target_based_coverage_file, f), original_coverage_file)


def file_compare(file1, file2):
    global ORIGINAL_COVERAGE_FILE
    global TARGET_BASED_COVERAGE_FILE
    if 'curve' not in file1 and '.ec' in file1:
        print(('TARGET_BASED_COVERAGE_FILE', TARGET_BASED_COVERAGE_FILE))
        print((os.path.join(TARGET_BASED_COVERAGE_FILE, file1)))
        created_time_1 = os.path.getctime(os.path.join(TARGET_BASED_COVERAGE_FILE, file1))
    else:
        print((os.path.join(ORIGINAL_COVERAGE_FILE, file1)))
        created_time_1 = os.path.getctime(os.path.join(ORIGINAL_COVERAGE_FILE, file1))
    if 'curve' not in file2 and '.ec' in file2:
        print((os.path.join(TARGET_BASED_COVERAGE_FILE, file2)))
        created_time_2 = os.path.getctime(os.path.join(TARGET_BASED_COVERAGE_FILE, file2))
    else:
        print((os.path.join(ORIGINAL_COVERAGE_FILE, file2)))
        created_time_2 = os.path.getctime(os.path.join(ORIGINAL_COVERAGE_FILE, file2))
    return int(created_time_1 * 100) - int(created_time_2 * 100)


def get_curve_data(original_coverage_file, new_coverage_file, app_source_file):
    global MAX_ID
    instruction_coverage_data = {}
    branch_coverage_data = {}
    cxty_coverage_data = {}
    line_coverage_data = {}
    method_coverage_data = {}
    class_coverage_data = {}
    activity_coverage_data = {}
    file_list = os.listdir(original_coverage_file)
    if not file_list:
        return
    file_list.sort(file_compare)
    for f in file_list:
        if 'curve' not in f and '.ec' in f:
            print(('file_name: ' + f + ' created time: ' + str(os.path.getctime(os.path.join(TARGET_BASED_COVERAGE_FILE, f)))))
        else:
            print(('file_name: ' + f + ' created time: ' + str(os.path.getctime(os.path.join(original_coverage_file, f)))))

    for f in file_list:
        if '.log' in f:
            continue
        if '.DS_Store' in f:
            continue
        if 'curve' not in f:
            shutil.copy2(os.path.join(original_coverage_file, f), new_coverage_file)
            continue
        else:
            shutil.copy2(os.path.join(original_coverage_file, f), new_coverage_file)
            generate_test_report(new_coverage_file, app_source_file)
            instruction_coverage, branch_coverage, cxty_coverage, line_coverage, method_coverage, class_coverage, activity_coverage = get_coverage_data(app_source_file)
            pattern = re.compile('coverage_curve_(\\S+).ec')
            file_id = re.findall(pattern, f)[0]
            instruction_coverage_data[int(file_id)] = instruction_coverage
            branch_coverage_data[int(file_id)] = branch_coverage
            cxty_coverage_data[int(file_id)] = cxty_coverage
            line_coverage_data[int(file_id)] = line_coverage
            method_coverage_data[int(file_id)] = method_coverage
            class_coverage_data[int(file_id)] = class_coverage
            activity_coverage_data[int(file_id)] = activity_coverage
            if int(file_id) > MAX_ID:
                MAX_ID = int(file_id)
            else:
                continue

    generate_test_report(new_coverage_file, app_source_file)
    instruction_coverage, branch_coverage, cxty_coverage, line_coverage, method_coverage, class_coverage, activity_coverage = get_coverage_data(app_source_file)
    instruction_coverage_data['final'] = instruction_coverage
    branch_coverage_data['final'] = branch_coverage
    cxty_coverage_data['final'] = cxty_coverage
    line_coverage_data['final'] = line_coverage
    method_coverage_data['final'] = method_coverage
    class_coverage_data['final'] = class_coverage
    activity_coverage_data['final'] = activity_coverage
    print(('data_id: final' + ' instruction_coverage:' + str(instruction_coverage)))
    print(('data_id: final' + ' branch_coverage:' + str(branch_coverage)))
    print(('data_id: final' + ' cxty_coverage:' + str(cxty_coverage)))
    print(('data_id: final' + ' line_coverage:' + str(line_coverage)))
    print(('data_id: final' + ' method_coverage:' + str(method_coverage)))
    print(('data_id: final' + ' class_coverage:' + str(class_coverage)))
    print(('data_id: final' + ' activity_coverage:' + str(activity_coverage)))
    return (instruction_coverage_data, branch_coverage_data, cxty_coverage_data, line_coverage_data, method_coverage_data, class_coverage_data, activity_coverage_data)


def generate_test_report(new_coverage_file, app_source_file):
    os.chdir(app_source_file)
    cmd = 'gradle jacocoTestReportMergeWithParameter -PnewCoverageFilePath=' + new_coverage_file
    os.system(cmd)


def get_coverage_data(app_source_file):
    global APP_SOURCE_FILE
    global MANIFEST_FILE
    instruction_coverage = 0.0
    branch_coverage = 0.0
    cxty_coverage = 0.0
    line_coverage = 0.0
    method_coverage = 0.0
    class_coverage = 0.0
    activity_coverage = 0.0
    file_path = app_source_file + '/build/reports/jacoco/jacocoTestReportMergeWithParameter' + '/jacocoTestReportMergeWithParameter.xml'
    if APP_SOURCE_FILE == '/home/tim/AndroidStudioProjects/Q-testing/runnerup':
        file_path = app_source_file + '/app/build/reports/jacoco/jacocoTestReportMergeWithParameter' + '/jacocoTestReportMergeWithParameter.xml'
    dom = xml.dom.minidom.parse(file_path)
    root = dom.documentElement
    children = root.childNodes
    for counter in children:
        if counter.getAttribute('type') == 'INSTRUCTION':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            instruction_coverage = float(covered) / float(missed + covered)
        if counter.getAttribute('type') == 'BRANCH':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            branch_coverage = float(covered) / float(missed + covered)
        if counter.getAttribute('type') == 'LINE':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            line_coverage = float(covered) / float(missed + covered)
        if counter.getAttribute('type') == 'COMPLEXITY':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            cxty_coverage = float(covered) / float(missed + covered)
        if counter.getAttribute('type') == 'METHOD':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            method_coverage = float(covered) / float(missed + covered)
        if counter.getAttribute('type') == 'CLASS':
            missed = int(counter.getAttribute('missed'))
            covered = int(counter.getAttribute('covered'))
            class_coverage = float(covered) / float(missed + covered)

    activity_names = []
    covered_activity_num = 0
    manifest_file = ''
    print(MANIFEST_FILE)
    if MANIFEST_FILE == '':
        manifest_file = app_source_file + '/app/src/main/AndroidManifest.xml'
    else:
        manifest_file = MANIFEST_FILE
    manifest_dom = xml.dom.minidom.parse(manifest_file)
    manifest_root = manifest_dom.documentElement
    application_node = manifest_root.getElementsByTagName('application')[0]
    activity_nodes = application_node.getElementsByTagName('activity')
    for activity_node in activity_nodes:
        name_str = activity_node.getAttribute('android:name')
        name_strs = name_str.split('.')
        activity_name = name_strs[len(name_strs) - 1]
        print(('activity name: ' + activity_name))
        activity_names.append(activity_name)

    package_nodes = root.getElementsByTagName('package')
    for package_node in package_nodes:
        class_nodes = package_node.getElementsByTagName('class')
        for class_node in class_nodes:
            class_name_str = class_node.getAttribute('name')
            name_strs = class_name_str.split('/')
            class_name = name_strs[len(name_strs) - 1]
            if class_name in activity_names:
                print(('class name: ' + class_name))
                for counter in class_node.childNodes:
                    if counter.getAttribute('type') == 'CLASS':
                        if int(counter.getAttribute('covered')) > 0:
                            print('covered')
                            covered_activity_num += 1
                        else:
                            print('not covered')

    activity_coverage = float(covered_activity_num) / float(len(activity_names))
    return (
     instruction_coverage, branch_coverage, cxty_coverage, line_coverage, method_coverage, class_coverage, activity_coverage)


def save_curve_data(instruction_coverage_data, branch_coverage_data, cxty_coverage_data, line_coverage_data, method_coverage_data, class_coverage_data, activity_coverage_data, new_coverage_file):
    f = open(new_coverage_file + '/curve_data_all.txt', 'w')
    sorted_instruction_data = collections.OrderedDict(instruction_coverage_data)
    sorted_branch_data = collections.OrderedDict(branch_coverage_data)
    sorted_cxty_data = collections.OrderedDict(cxty_coverage_data)
    sorted_line_data = collections.OrderedDict(line_coverage_data)
    sorted_method_data = collections.OrderedDict(method_coverage_data)
    sorted_class_data = collections.OrderedDict(class_coverage_data)
    sorted_activity_data = collections.OrderedDict(activity_coverage_data)
    for data_id in list(sorted_instruction_data.keys()):
        if data_id == 'final':
            f.write('time: final\t' + 'instruction_coverage: ' + str(round(sorted_instruction_data[data_id], 4)) + '\t')
            f.write('branch_coverage: ' + str(round(sorted_branch_data[data_id], 4)) + '\t')
            f.write('cxty_coverage: ' + str(round(sorted_cxty_data[data_id], 4)) + '\t')
            f.write('line_coverage: ' + str(round(sorted_line_data[data_id], 4)) + '\t')
            f.write('method_coverage: ' + str(round(sorted_method_data[data_id], 4)) + '\t')
            f.write('class_coverage: ' + str(round(sorted_class_data[data_id], 4)) + '\t')
            f.write('activity_coverage: ' + str(round(sorted_activity_data[data_id], 4)) + '\n')
        else:
            f.write('time: ' + str(int(data_id) * TIME_INTERVAL) + '\t' + 'instruction_coverage: ' + str(round(sorted_instruction_data[data_id], 4)) + '\t')
            f.write('branch_coverage: ' + str(round(sorted_branch_data[data_id], 4)) + '\t')
            f.write('cxty_coverage: ' + str(round(sorted_cxty_data[data_id], 4)) + '\t')
            f.write('line_coverage: ' + str(round(sorted_line_data[data_id], 4)) + '\t')
            f.write('method_coverage: ' + str(round(sorted_method_data[data_id], 4)) + '\t')
            f.write('class_coverage: ' + str(round(sorted_class_data[data_id], 4)) + '\t')
            f.write('activity_coverage: ' + str(round(sorted_activity_data[data_id], 4)) + '\n')

    f.write('===========================\n')
    f.write('30min:\t' + 'instruction_coverage: ' + str(round(sorted_instruction_data[MAX_ID / 2], 4)) + '\t')
    f.write('branch_coverage: ' + str(round(sorted_branch_data[MAX_ID / 2], 4)) + '\t')
    f.write('cxty_coverage: ' + str(round(sorted_cxty_data[MAX_ID / 2], 4)) + '\t')
    f.write('line_coverage: ' + str(round(sorted_line_data[MAX_ID / 2], 4)) + '\t')
    f.write('method_coverage: ' + str(round(sorted_method_data[MAX_ID / 2], 4)) + '\t')
    f.write('class_coverage: ' + str(round(sorted_class_data[MAX_ID / 2], 4)) + '\t')
    f.write('activity_coverage: ' + str(round(sorted_activity_data[MAX_ID / 2], 4)) + '\n')
    f.write('60min:\t' + 'instruction_coverage: ' + str(round(sorted_instruction_data['final'], 4)) + '\t')
    f.write('branch_coverage: ' + str(round(sorted_branch_data['final'], 4)) + '\t')
    f.write('cxty_coverage: ' + str(round(sorted_cxty_data['final'], 4)) + '\t')
    f.write('line_coverage: ' + str(round(sorted_line_data['final'], 4)) + '\t')
    f.write('method_coverage: ' + str(round(sorted_method_data['final'], 4)) + '\t')
    f.write('class_coverage: ' + str(round(sorted_class_data['final'], 4)) + '\t')
    f.write('activity_coverage: ' + str(round(sorted_activity_data['final'], 4)) + '\n')
    f.close()


def calculate_coverage():
    global APP_SOURCE_FILE
    global Benchmark
    global MANIFEST_FILE
    global NEW_COVERAGE_FILE
    global ORIGINAL_COVERAGE_FILE
    global TARGET_BASED_COVERAGE_FILE
    bundle_dir = getattr(sys, '_MEIPASS', os.path.abspath(os.path.dirname(__file__)))
    path_to_config = os.path.abspath(os.path.join(bundle_dir, './Config.txt'))
    config = configparser.ConfigParser()
    config.read(path_to_config)
    APP_SOURCE_FILE = config.get('Path', 'APP_SOURCE_PATH')
    APK_NAME = config.get('Path', 'APK_NAME')
    Benchmark = config.get('Path', 'Benchmark')
    if not Benchmark.endswith('/'):
        Benchmark = Benchmark + '/'
    MANIFEST_FILE = config.get('Path', 'MANIFEST_FILE')
    TARGET_BASED_COVERAGE_FILE = Benchmark + APK_NAME.split('.')[0] + '-output-1/coverage_merge_1'
    ORIGINAL_COVERAGE_FILE = Benchmark + APK_NAME.split('.')[0] + '-output-1/coverage_original_1'
    NEW_COVERAGE_FILE = Benchmark + APK_NAME.split('.')[0] + '-output-1/coverage_new_1'
    copy_restart_coverage_file(TARGET_BASED_COVERAGE_FILE, ORIGINAL_COVERAGE_FILE)
    instruction_data, branch_data, cxty_data, line_data, method_data, class_data, activity_data = get_curve_data(ORIGINAL_COVERAGE_FILE, NEW_COVERAGE_FILE, APP_SOURCE_FILE)
    save_curve_data(instruction_data, branch_data, cxty_data, line_data, method_data, class_data, activity_data, NEW_COVERAGE_FILE)


if __name__ == '__main__':
    calculate_coverage()
