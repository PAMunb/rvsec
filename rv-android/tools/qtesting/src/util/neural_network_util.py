import re
from tensorflow.keras import backend as K
from tensorflow.keras.layers import Layer
from tensorflow.keras.preprocessing.sequence import pad_sequences
import numpy as np, itertools
max_length = 0

def string_to_vector(df):
    global max_length
    for index, row in df.iterrows():
        for gui_hierarchy in ['gui_hierarchy1', 'gui_hierarchy2']:
            s2v = []
            for vector_string in str(row[gui_hierarchy]).split('\n'):
                if vector_string != '':
                    vector = []
                    for sub_string in vector_string.split(' '):
                        if sub_string != '':
                            vector.append(float(sub_string))

                    s2v.append(vector)

            df.at[(index, gui_hierarchy + '_v')] = s2v
            if len(s2v) > max_length:
                max_length = len(s2v)
                print(('max length is now : ' + str(max_length)))

    return df


def split_and_zero_padding(df, max_seq_length):
    X = {'left': (df['gui_hierarchy1_v']), 'right': (df['gui_hierarchy2_v'])}
    for dataset, side in itertools.product([X], ['left', 'right']):
        dataset[side] = pad_sequences(dataset[side], padding='pre', truncating='post', maxlen=max_seq_length)

    return dataset


class ManDist(Layer):
    """
    Keras Custom Layer that calculates Manhattan Distance.
    """

    def __init__(self, **kwargs):
        self.result = None
        super(ManDist, self).__init__(**kwargs)
        return

    def build(self, input_shape):
        print(input_shape)
        super(ManDist, self).build(input_shape)

    def call(self, x, **kwargs):
        self.result = K.exp(-K.sum(K.abs(x[0] - x[1]), axis=1, keepdims=True))
        return self.result

    def compute_output_shape(self, input_shape):
        return K.int_shape(self.result)

    def get_config(self):
        base_config = super(ManDist, self).get_config()
        return base_config


class EmptyWord2Vec:
    """
    Just for test use.
    """
    vocab = {}
    word_vec = {}
