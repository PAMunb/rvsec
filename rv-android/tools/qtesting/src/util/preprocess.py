from hashlib import md5
import logging, os, sys, xml.dom.minidom, xml.etree.ElementTree, subprocess
from tensorflow.keras.preprocessing.sequence import pad_sequences
logging.basicConfig(format='%(asctime)s : %(filename)s[line:%(lineno)d] : %(levelname)s : %(message)s', level=logging.INFO)
bundle_dir = getattr(sys, '_MEIPASS', os.path.abspath(os.path.dirname(__file__)))
WIDGET_CLASS_FILE_PATH = os.path.abspath(os.path.join(bundle_dir, './widget_classes.txt'))
SCREEN_RESOLUTION_X = 768
SCREEN_RESOLUTION_Y = 1280

class Preprocess(object):

    def __init__(self, DEVICE_ID):
        global SCREEN_RESOLUTION_X
        global SCREEN_RESOLUTION_Y
        self.widget_class_map = {}
        self.widget_class_list = []
        self.read_widget_classes()
        result = subprocess.getoutput('adb -s ' + DEVICE_ID + ' shell wm size')
        if 'Physical size:' in result:
            offset = result.rindex('Physical size:') + 14
            xAndyStr = result[offset:]
            xAndy = xAndyStr.split('x')
            SCREEN_RESOLUTION_X = int(xAndy[0])
            SCREEN_RESOLUTION_Y = int(xAndy[1])

    def read_widget_classes(self):
        f = open(WIDGET_CLASS_FILE_PATH, 'r')
        for line in f.readlines():
            self.widget_class_list.append(line.strip('\n'))

    def extract_vectors(self, xml_tree):
        vectors = []
        root = xml_tree.getroot()
        self.preorder_traverse(vectors, root)
        vectors = self.split_and_zero_padding([vectors], 100)
        return vectors

    def preorder_traverse(self, vectors, node):
        if 'hierarchy' != node.tag:
            vector = self.extract_vector_from_node(node)
            vectors.append(vector)
        for child in node:
            self.preorder_traverse(vectors, child)

    def extract_vector_from_node(self, node):
        node_vector = [0.0] * (5 + len(self.widget_class_list) + 1 + 160)
        if node.get('NAF') == '' or not node.get('NAF'):
            node_vector[0] = 0.0
        elif node.get('NAF') == 'true':
            node_vector[0] = 1.0
        bounds_string = node.get('bounds')
        if bounds_string == '' or not bounds_string:
            print(node)
            node_vector[1] = 0.0
            node_vector[2] = 0.0
            node_vector[3] = 0.0
            node_vector[4] = 0.0
        else:
            lt_bounds_string = bounds_string[0:bounds_string.rfind('[')]
            rb_bounds_string = bounds_string[bounds_string.rfind('['):len(bounds_string)]
            lt_x = int(lt_bounds_string[1:lt_bounds_string.find(',')])
            lt_y = int(lt_bounds_string[lt_bounds_string.find(',') + 1:lt_bounds_string.rfind(']')])
            rb_x = int(rb_bounds_string[1:rb_bounds_string.find(',')])
            rb_y = int(rb_bounds_string[rb_bounds_string.find(',') + 1:rb_bounds_string.rfind(']')])
            node_vector[1] = round(lt_x / SCREEN_RESOLUTION_X, 2)
            node_vector[2] = round(lt_y / SCREEN_RESOLUTION_Y, 2)
            node_vector[3] = round(rb_x / SCREEN_RESOLUTION_X, 2)
            node_vector[4] = round(rb_y / SCREEN_RESOLUTION_Y, 2)
        widget_class = node.get('class')
        one_hot_num = 0
        for i in range(0, len(self.widget_class_list)):
            if self.widget_class_list[i] == widget_class:
                one_hot_num = i
                break

        if widget_class == self.widget_class_list[one_hot_num]:
            node_vector[one_hot_num + 5] = 1.0
        else:
            node_vector[len(self.widget_class_list) + 5] = 1.0
        content_desc = node.get('content-desc')
        md5_str_16 = str(md5(content_desc.encode('utf-8')).hexdigest())
        md5_str_10 = str(int(md5_str_16, 16))
        if len(md5_str_10) > 40:
            logging.error('length is beyond 40')
        for i in range(0, len(md5_str_10)):
            node_vector[len(self.widget_class_list) + 6 + i] = float(md5_str_10[i]) / 10
            if i >= 39:
                break

        package = node.get('package')
        md5_str_16 = str(md5(package.encode('utf-8')).hexdigest())
        md5_str_10 = str(int(md5_str_16, 16))
        if len(md5_str_10) > 40:
            logging.error('length is beyond 40')
        for i in range(0, len(md5_str_10)):
            node_vector[len(self.widget_class_list) + 46 + i] = float(md5_str_10[i]) / 10
            if i >= 39:
                break

        resource_id = node.get('resource-id')
        md5_str_16 = str(md5(resource_id.encode('utf-8')).hexdigest())
        md5_str_10 = str(int(md5_str_16, 16))
        if len(md5_str_10) > 40:
            logging.error('length is beyond 40')
        for i in range(0, len(md5_str_10)):
            node_vector[len(self.widget_class_list) + 86 + i] = float(md5_str_10[i]) / 10
            if i >= 39:
                break

        text = node.get('text')
        md5_str_16 = str(md5(text.encode('utf-8')).hexdigest())
        md5_str_10 = str(int(md5_str_16, 16))
        if len(md5_str_10) > 40:
            logging.error('length is beyond 40')
        for i in range(0, len(md5_str_10)):
            node_vector[len(self.widget_class_list) + 126 + i] = float(md5_str_10[i]) / 10
            if i >= 39:
                break

        return node_vector

    def split_and_zero_padding(self, df, max_seq_length):
        return pad_sequences(df, padding='pre', truncating='post', maxlen=max_seq_length, dtype='float', value=0.0)
