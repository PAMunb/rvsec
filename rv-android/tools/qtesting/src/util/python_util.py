def parse_action_string(action_string):
    if '@' in action_string:
        loc = action_string.index('@')
    else:
        loc = None
    if ':' in action_string:
        loc_type = action_string.index(':')
    else:
        loc_type = None
    if ')' in action_string:
        loc_scope = action_string.rindex(')')
    else:
        loc_scope = None
    action_id = action_string[0:loc]
    if loc_type is not None and loc_scope is not None and loc_scope < loc_type:
        action_cmd = action_string[loc + 1:loc_type]
        view_string = action_string[loc_type + 1:len(action_string)]
        print(('view_string is: ' + view_string))
        loc_view_text = view_string.index('@')
        if loc_view_text is not None:
            view_type = view_string[0:loc_view_text]
            view_text = view_string[loc_view_text + 2:len(view_string) - 2]
        else:
            view_type = view_string
            view_text = ''
    else:
        action_cmd = action_string[loc + 1:len(action_string)]
        view_type = ''
        view_text = ''
    return (
     action_id, action_cmd, view_type, view_text)
