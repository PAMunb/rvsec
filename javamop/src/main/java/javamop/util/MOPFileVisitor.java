package javamop.util;

import javamop.JavaMOPMain;

import java.io.IOException;
import java.nio.file.*;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.ArrayList;

import static java.nio.file.FileVisitResult.CONTINUE;

/**
 * Created by hx312 on 25/04/2015.
 */
public class MOPFileVisitor extends SimpleFileVisitor<Path> {
    private Path inputDirPath = null;
    private final Path outputDirPath;

    public MOPFileVisitor(Path outputDirPath) {
        this.outputDirPath = outputDirPath;
    }

    @Override
    public FileVisitResult preVisitDirectory(final Path dir,
                                   final BasicFileAttributes attributes)
        throws IOException {
        if (this.inputDirPath == null) {
            this.inputDirPath = dir;
        } else {
            Files.createDirectories(this.outputDirPath.resolve
                    (this.inputDirPath.relativize(dir)));
        }
        return CONTINUE;
    }

    @Override
    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
        String fileName = String.valueOf(file.getFileName());
        if (fileName.endsWith(".mop")) {
            //when the mop file is visited, its corresponding folder
            // must have already been created in the output directory
            // by the above preVisitDirectory method.
            Path outputPath = this.outputDirPath.resolve
                    (this.inputDirPath.relativize(file)).getParent();
            JavaMOPMain.main(new String[]{file.toFile().getAbsolutePath(),
                            "-d", outputPath.toFile().getAbsolutePath()});
        }
        return CONTINUE;
    }

}
