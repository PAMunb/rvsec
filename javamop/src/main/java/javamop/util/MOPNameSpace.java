// Copyright (c) 2002-2014 JavaMOP Team. All Rights Reserved.
package javamop.util;


import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

/**
 * A namespace class to keep track of which variable names are being used to avoid duplicate
 * variables with the same name.
 */
public final class MOPNameSpace {
    
    /**
     * Private to prevent instantiation.
     */
    private MOPNameSpace() {
        
    }
    
    static private boolean used = false;
    static private final List<String> userVariables = new ArrayList<String>();
    static private final List<String> mopVariables = new ArrayList<String>();
    static private final Map<String, String> mapVars = new HashMap<String, String>();
    
    static private final HashSet<String> keywords = new HashSet<String>();
    
    static {
        keywords.add("this");
        keywords.add("null");
        keywords.add("MOP_lastevent");
        keywords.add("MOP_terminated");
        keywords.add("thisJoinPoint");
        
    }
    
    /**
     * Reinitialize the static data.
     */
    static public void init(){
        used = false;
    }
    
    /**
     * Register a new user variable.
     * @param varName The name of the user variable.
     * @throws javamop.util.MOPException If the name is reserved in JavaMOP.
     */
    static public void addUserVariable(final String varName) throws MOPException {
        if (used)
            throw new MOPException("Cannot update MOPNameSpace after once used");
        
        if(keywords.contains(varName))
            throw new MOPException(varName + " is reserved in JavaMOP/AspectJ. Please rename it.");
        
        if(!userVariables.contains(varName))
            userVariables.add(varName);
    }
    
    /**
     * Add a collection of user variables.
     * @param varNames A collection of user variable names.
     * @throws MOPException If any of the names is reserved in JavaMOP.
     */
    static public void addUserVariables(final Collection<String> varNames) throws MOPException {
        for (String varName : varNames)
            addUserVariable(varName);
    }
    
    /**
     * Check the existence of a named user variable.
     * @param varName The name of the variable to look for.
     * @return Whether there exists a user variable by that name or not.
     */
    static public boolean checkUserVariable(final String varName) {
        return userVariables.contains(varName);
    }
    
    /**
     * Retrieve a unique variable name with a given base for use in JavaMOP code.
     * @param varName The base of the variable name.
     * @return A unique name to use at a particular instnace.
     */
    static public String getMOPVar(final String varName) {
        used = true;
        
        String cachedVar = mapVars.get(varName);
        if (cachedVar != null)
            return cachedVar;
        
        if(keywords.contains(varName))
            return varName;
        
        int extNum = 1;
        if (userVariables.contains(varName) || mopVariables.contains(varName)) {
            while (userVariables.contains(varName + "_" + extNum) || 
                    mopVariables.contains(varName + "_" + extNum))
                extNum++;
            mapVars.put(varName, varName + "_" + extNum);
            mopVariables.add(varName + "_" + extNum);
            return varName + "_" + extNum;
        } else {
            mapVars.put(varName, varName);
            return varName;
        }
    }
}
