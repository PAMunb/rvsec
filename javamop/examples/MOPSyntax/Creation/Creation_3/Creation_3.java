// Copyright (c) 2002-2014 JavaMOP Team. All Rights Reserved.

import java.util.*;

public class Creation_3 {
    public int temp = 0;
    public void fun1(){
        temp = 1;
    }
    public void fun2(){
        temp = 2;
    }
    
    public static void main(String[] args){
        Creation_3 o = new Creation_3();
        
        System.out.println("fun2");
        o.fun2();
        System.out.println("fun1");
        o.fun1();
        System.out.println("fun2");
        o.fun2();
        System.out.println("fun1");
        o.fun1();
        
        System.out.println("main end");
    }
}
