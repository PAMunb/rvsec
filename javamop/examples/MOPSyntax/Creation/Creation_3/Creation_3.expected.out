fun2
fun1
fun2
fun1
main end
match
